src/tpm2_param_enc.c

```diff
@@ -0,0 +1,282 @@
+#include <wolftpm/tpm2_packet.h>
+
+#include <wolfssl/wolfcrypt/hmac.h>
+
+/* Routines for performing TPM Parameter Encryption
+ *
+ * NB: Only TPM2B_DATA parameters can be encrypted
+ *
+ * Only the first parameter of a TPM commmand can be encrypted.
+ * For example, the password auth of a TPM key. The encryption
+ * of command response and request are separate. There can be a
+ * communication exchange between the TPM and a client program
+ * where only the parameter in the request command is encrypted.
+ *
+ * This behavior depends on the sessionAttributes:
+ * - TPMA_SESSION_encrypt for command request
+ * - TPMA_SESSION_decrypt for command response
+ * Either one can be set separately or both can be set in one
+ * authorization session. This is up to the user(developer).
+ *
+ */
+
+/******************************************************************************/
+/* --- Local Functions -- */
+/******************************************************************************/
+
+/* This function performs key generation according to Part 1 of the TPM spec
+ * and returns the number of bytes generated, which may be zero.
+ *
+ * 'key' and 'keyStream' are pointers and they can not be NULL.
+ * Any other pointer arguments may be NULL.
+ * 'sizeInBits' must be no larger than (2^18)-1 = 256K bits (32385 bytes).
+ *
+ * Note: The "once" parameter is set to allow incremental generation of a large
+ * value. If this flag is TRUE, "sizeInBits" is used in the HMAC computation
+ * but only one iteration of the KDF is performed. This would be used for
+ * XOR obfuscation so that the mask value can be generated in digest-sized
+ * chunks rather than having to be generated all at once in an arbitrarily
+ * large buffer and then XORed into the result. If "once" is TRUE, then
+ * "sizeInBits" must be a multiple of 8.
+ *
+ * Any error in the processing of this command is considered fatal.
+ *
+ * Return values:
+ *     0    hash algorithm is not supported or is TPM_ALG_NULL
+ *    >0    the number of bytes in the 'keyStream' buffer
+ *
+ */
+static int TPM2_KDFa(
+    TPM_ALG_ID  hashAlg,    /* IN: hash algorithm used in HMAC */
+    TPM2B_DATA  *key,       /* IN: key, can be NULL if authValue is used */
+    const char  *label,     /* IN: a 0-byte terminated label used in KDF */
+    TPM2B_DATA  *contextU,  /* IN: context U */
+    TPM2B_DATA  *contextV,  /* IN: context V */
+    word32      sizeInBits, /* IN: size of generated key in bits */
+    byte        *keyStream, /* OUT: key buffer */
+    word32      *counterInc,/* IN/OUT: See Note for incremental operations */
+    int         doOnce      /* IN: TRUE if only one iteration is performed */
+)
+{
+    int ret, hashType;
+    Hmac hmac;
+    word32 counter = 0;
+    int hLen, outLen, lLen = 0;
+    byte *outStream;
+    byte uint32Buf[sizeof(UINT32)];
+
+    if (key == NULL || keyStream == NULL)
+        return BAD_FUNC_ARG;
+
+    if (doOnce != FALSE && (sizeInBits & 7) != 0)
```

Comment:
Let's avoid using `FALSE` and `TRUE`. Stick with 0 and 1. If you want you can stick with the TPM's def of it "YES" and "NO" in tpm2_types.h.

Reply:
OK